src/schedule/zephyr_dp_schedule.c

```diff
@@ -0,0 +1,358 @@
+	k_spinlock_key_t lock_key;
+
+	if (cpu_get_id() != PLATFORM_PRIMARY_CORE_ID)
+		return;
+
+	if (!dp_sch)
+		return;
+
+	lock_key = scheduler_dp_lock();
+	list_for_item(tlist, &dp_sch->tasks) {
+		curr_task = container_of(tlist, struct task, list);
+		pdata = curr_task->priv_data;
+
+		if (pdata->ticks_to_trigger == 0) {
+			if (curr_task->state == SOF_TASK_STATE_QUEUED) {
+				/* set new trigger time, start the thread */
+				pdata->ticks_to_trigger = pdata->ticks_period;
+				curr_task->state = SOF_TASK_STATE_RUNNING;
+				k_sem_give(&pdata->sem);
+			}
+		} else {
+			if (curr_task->state == SOF_TASK_STATE_QUEUED ||
+			    curr_task->state == SOF_TASK_STATE_RUNNING)
+				/* decrease num of ticks to re-schedule */
+				pdata->ticks_to_trigger--;
+		}
+	}
+	scheduler_dp_unlock(lock_key);
+}
+
+static int scheduler_dp_task_cancel(void *data, struct task *task)
+{
+	(void)(data);
+	k_spinlock_key_t lock_key;
+
+	/* this is asyn cancel - mark the task as canceled and remove it from scheduling */
+	lock_key = scheduler_dp_lock();
+
+	task->state = SOF_TASK_STATE_CANCEL;
+	list_item_del(&task->list);
+
+	scheduler_dp_unlock(lock_key);
+
+	return 0;
+}
+
+static int scheduler_dp_task_free(void *data, struct task *task)
+{
+	k_spinlock_key_t lock_key;
+	struct task_dp_pdata *pdata = task->priv_data;
+
+	/* abort the execution of the thread */
+	k_thread_abort(pdata->thread_id);
+
+	lock_key = scheduler_dp_lock();
+	list_item_del(&task->list);
+	task->priv_data = NULL;
+	task->state = SOF_TASK_STATE_FREE;
+	scheduler_dp_unlock(lock_key);
+
+	/* free all allocated resources */
+	rfree(pdata->p_stack);
+	rfree(pdata->thread_id);
+	rfree(pdata);
+
+	return 0;
+}
+
+/* Thread function called in component context, on target core */
+static void dp_thread_fn(void *p1, void *p2, void *p3)
+{
+	struct task *task = p1;
+	(void) p2;
+	(void) p3;
+	struct task_dp_pdata *task_pdata = task->priv_data;
+	k_spinlock_key_t lock_key;
+	enum task_state state;
+
+	while (1) {
+		/*
+		 * the thread is started immediately after creation, it will stop on semaphore
+		 * Semaphore will be released once the task is ready to process
+		 */
+		k_sem_take(&task_pdata->sem, K_FOREVER);
+
+		if (task->state == SOF_TASK_STATE_RUNNING)
+			state = task_run(task);
+		else
+			state = task->state;	/* to avoid undefined variable warning */
+
+		lock_key = scheduler_dp_lock();
+		/*
+		 * check if task is still running, may have been canceled by external call
+		 * if not, set the state returned by run procedure
+		 */
+		if (task->state == SOF_TASK_STATE_RUNNING) {
+			task->state = state;
+			switch (state) {
+			case SOF_TASK_STATE_RESCHEDULE:
+				/* mark to reschedule, schedule time is already calculated */
+				task->state = SOF_TASK_STATE_QUEUED;
+				break;
+
+			case SOF_TASK_STATE_CANCEL:
+			case SOF_TASK_STATE_COMPLETED:
+				/* remove from scheduling */
+				list_item_del(&task->list);
+				break;
+
+			default:
+				/* illegal state, serious defect, won't happen */
```

Comment:
If it's a "will not happen" case, might be preferable to use an __ASSERT() here instead.

Reply:
I just changed from assert ;)